### ISSUE TITLE
Add capability to restrict flag values to a set of allowed values

### ISSUE THREAD

**Issue by ian-howell** [spf13/pflag#236]
I have a use-case where I need to perform validation on flag values. In other words, I need to check that the value passed in by the user belongs to some set of valid values. I end up writing a lot of code that looks like this:

```
package main

import (
	"fmt"
	"os"

	flag "github.com/spf13/pflag"
)

var color string

func main() {
	addFlags()
	flag.Parse()
	if err := validateFlags(); err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v", err)
		os.Exit(1)
	}
	fmt.Printf("Selected color is %s\n", color)
}

func addFlags() {
	flag.StringVar(&color, "color", "", "The color for the example")
}

func validateFlags() error {
	validColors := []string{"red", "blue", "yellow"}
	for _, validColor := range validColors {
		if color == validColor {
			// color is valid
			return nil
		}
	}
	// if we're here, color is invalid
	return fmt.Errorf("Value '%s' is invalid for flag 'color'. Valid values "+
		"come from the set %v", color, validColors)
}
```

I'd like to be able to do something like the following:

```
	flag.StringVarRestricted(
		&color,                            // pointer to variable
		"color",                           // flag name
		"",                                // default value
		"The color for the example",       // description
		[]string{"red", "blue", "yellow"}) // Allowed values
```

I wouldn't mind taking a shot at implementing something like this if it sounds acceptable.

**Comment by macintacos** [comment]
Just dropping a comment noting that this is the most thumbs-upped issue in this repo, and I'd love to see it come to fruition 🙂

Right now I do some reflection-gymnastics based on pre-defined types to validate and define flags, so if this were added in this project it'd obviate the need for that custom logic and make my codebase that much easier to maintain. I see #237 was opened, but looks like it didn't get much traction?

**Comment by cornfeedhobo** [comment]
@macintacos I'll take a swing at implementing something like this on [my fork](https://github.com/cornfeedhobo/pflag).

**Comment by d3an** [comment]
Python's [argparse](https://docs.python.org/3/library/argparse.html#choices) supports flags with predefined choices. I want to replicate the same behaviour in Cobra, and think it would be a great addition.

**Comment by migueleliasweb** [comment]
This is would be so useful. Is there anything keeping this from getting merged? 🙏

**Comment by felipe-avelar** [comment]
I think this feature would be awesome, I would like to suggest a change, instead a slice of valid values could be a bool function that validates the value 😃

**Comment by pmenglund** [comment]
@felipe-avelar  that is a better approach as it is more flexiable, so using the initial example it would become:
```
	flag.StringVarValidated(
		&color,                            // pointer to variable
		"color",                           // flag name
		"",                                // default value
		"The color for the example",       // description
		func (s string) bool { ... }(),    // Function to evaluate valid values
```

**Comment by fastcat** [comment]
Alternative POV: If you provide a slice of valid values, then help output can tell the user what they are, and shell autocomplete integration can leverage that information.

Of course, not all cases can be handled by a slice, so this is probably a case of "yes, and". For example, could have the validation function, and an (optional) slice of "suggested" values?

**Comment by cornfeedhobo** [comment]
@fastcat You make a great point.

@pmenglund I like your suggestion and think it should be a feature either way..

I will try to find time soon to implement this on my fork, which also has other features, and works with cobra (instructions in the readme).

**Comment by thalesfsp** [comment]
Is this project dead? How come this issue is still open and stale? Where are the repo admins?

**Comment by thalesfsp** [comment]
Is this project dead? How come this issue is still open and stale? Where are the repo admins?
cc: @spf13

**Comment by pmenglund** [comment]
@thalesfsp last I check this is an open source project, so you either avoid looking a gift horse in the mouth and accept the free software with its limitations, or you implement the missing feature you want on your own (or pay someone to do it for you).

**Comment by cornfeedhobo** [comment]
> @thalesfsp last I check this is an open source project, so you either avoid looking a gift horse in the mouth and accept the free software with its limitations, or you implement the missing feature you want on your own (or pay someone to do it for you).

Why is this snarky reply so popular? The question is completely valid given the neglect this library sees and the outright hostility everyone encounters when touching the subject on the cobra repos.

A fork has been floated on so many issues with this repo and everyone keeps shooting it down and complaining when people mention it, so please don't act like that's a real solution for the average user. If you have any experience in the industry, you know that's a straw argument.

**Comment by pmenglund** [comment]
It is popular because so many put demands on people who devote their free time to maintaining open source projects, with no compensation but the "fame". My guess is that @spf13 has a demanding day-job, and probably a family, so any sliver of spare time is probably spent on something to unwind.

So what are the alternatives? You can whine or you can do something to fix it. Whining gets you snarky comments from people who have maintained open source projects.

That said, I should not have posted the snarky comment. Instead I should have offered advice, but my patience quota had run out.

Now it has been refilled, so here goes:

Go makes it easy very to fork this repo and use a [replace](https://golang.org/ref/mod#go-mod-file-replace) directive in your `go.mod` to swap it out for your own fork, where you merged #237. That is a solution the average user can use. Not optimal but workable.

Since @cornfeedhobo has forked it, @thalesfsp can just add this to `go.mod` and get all benefits from the fork:
```
replace (
    github.com/spf13/pflag => github.com/cornfeedhobo/pflag
)
```

**Comment by cornfeedhobo** [comment]
@pmenglund thanks for such a thoughtful reply. Re-reading my comment and I can see I was more incensed than I meant to be. Thanks for reading through that, and my apologies.

**Comment by frankywahl** [comment]
I think this is achievable with custom flags (type), as flag is only an interface: 

Sample code: 

```go
package main

import (
	"fmt"
	"strings"

	flag "github.com/spf13/pflag"
)

type enum struct {
	Allowed []string
	Value   string
}

// newEnum give a list of allowed flag parameters, where the second argument is the default
func newEnum(allowed []string, d string) *enum {
	return &enum{
		Allowed: allowed,
		Value:   d,
	}
}

func (a enum) String() string {
	return a.Value
}

func (a *enum) Set(p string) error {
	isIncluded := func(opts []string, val string) bool {
		for _, opt := range opts {
			if val == opt {
				return true
			}
		}
		return false
	}
	if !isIncluded(a.Allowed, p) {
		return fmt.Errorf("%s is not included in %s", p, strings.Join(a.Allowed, ","))
	}
	a.Value = p
	return nil
}

func (a *enum) Type() string {
	return "string"
}

// Main program starts here

func main() {
	color := newEnum([]string{"red", "green", "blue"}, "green")
	flag.VarP(color, "color", "c", "the color you want")
	flag.Parse()
	fmt.Printf("You have selected %s\n", color)
}
```

**Comment by gingerwizard** [comment]
@frankywahl is the best answer here. Trivial to extend and presumably exactly what the author intended.

**Comment by TheDiveO** [comment]
While [enumflag](https://github.com/TheDiveO/enumflag) is not a generic add-on for spf13, it covers the use case of enumeration flags (and slices). Maybe that might be of some help in such cases?

**Comment by memreflect** [comment]
The addition of `Func/FuncP` in #350 (disclaimer: i'm the PR author) would also fix the issue in a more generic (and indirect) manner in the case of only one or two flags:

```go
package main

import (
	"fmt"
	"os"
	"sort"
	"strings"

	flag "github.com/spf13/pflag"
)

func main() {
	color := ""
	flag.Func("color", "use `COLOR` for the example, or \"help\"",
		func(s string) error {
			// NOTE: keep sorted!
			validColors := []string{"blue", "red", "yellow"}

			if s == "help" {
				// special case: --color=help
				fmt.Fprintln(flag.CommandLine.Output(), "Valid colors:", strings.Join(validColors, ", "))
				os.Exit(2) // "invalid argument" error if `return flag.ErrHelp` is used :(
			}

			// sort.SearchStrings can be used since validColors is already sorted.
			i := sort.SearchStrings(validColors, s)
			if i < len(validColors) && validColors[i] == s {
				color = s
				return nil
			}
			return fmt.Errorf("invalid color %q; use --color=help to list valid colors", s)
		},
	)
	flag.Parse()
	if color == "" {
		flag.Lookup("color").Value.Set("") // stay DRY
	} else {
		fmt.Printf("Selected color is %s\n", color)
	}
}
```

For many such flags, however, [a custom `Var` implementation as suggested](https://github.com/spf13/pflag/issues/236#issuecomment-931600452) by @frankywahl is a much better solution, though i would sort `enum.Allowed` before returning from `newEnum` to allow faster searches with `sort.SearchStrings`.
Of course, that is merely the most straightforward implementation; there may be better alternatives to consider.

Naturally you'll have people asking if one can extend this to yet another `Var` implementation with a one-to-one mapping from a restricted value to a custom function, or a way to restrict a range of `int` values without storing all possible `int` values in `enum.Allowed`, or some other request to add more code to pflag just so their own code does not need to do it.
I would not be surprised if an author/maintainer/contributor of pflag is especially wary of adding such `Var` implementations since pflag already has significantly more `Var` implementations than Go's standard flag package, even excluding the `P` variants:

```none
bash$ topfuncs() { go doc $1 | grep '^func.*name.*usage' | grep -v '^func[^(]*P' ; }
bash$ wc -l <(topfuncs flag) <(topfuncs github.com/spf13/pflag)
      19 /tmp//sh-np.azCogP
      65 /tmp//sh-np.MqaD5O
      84 total
```

That said, i agree an implementation of `Var` that allows you to restrict flag values is general enough and popular enough to be a useful addition with the caveat that `enum` is a bad name as various proposals exist to add typed enums to Go, meaning `enum` might become a keyword or a built-in at some point in the future.

**Comment by jerome-diver** [comment]
Hi, an ther one solution i choosed can be to add a PersistentPreRunE qualifier to your root command to operate on all general flags, or on the command linked flags concerned by.
Then you can check for the variable that has catched the value and operate a regex or a switch/case or what ever you want.
This way should add a "clean" solution (i mean... readable and splitable easy in a MVC design too)  to contourn the lake of time interest from the owner of this nice repo.
If you need an example, please ask, i will modify my comment to show you that (but i think you got the point there).

**Comment by tomasaschan** [comment]
This issue has not been updated in a long time. Is it still relevant?

(If I receive no answer before the end of the summer, I'll close it. Feel free to re-open if you come here after that happens and it is relevant to you - but if you do, please also comment and specify what outcome you're interested in!)

**Comment by rgmz** [comment]
> This issue has not been updated in a long time. Is it still relevant?

I _believe_ so. To my knowledge, there's no builtin capability to restrict a flag to a set of allowed values. Several workarounds and alternatives have been shared, however, it would be nice if there was out-of-the-box support — or a definitive 'no, this won't be added, use _x_ instead'.

Edit: personally, I would love for this to capability to be added to pflag.

**Comment by macintacos** [comment]
I would definitely still like to see some version of this implemented.

**Comment by tomasaschan** [comment]
@rgmz @macintacos The comment above was scripted, targeting all issues that haven't seen updates since 2023 - but now that I read it properly I'm inclined to say that this feature is out of scope for pflag. That doesn't mean it's not a valuable feature - but it's already built into Cobra, and given the close relationship between the two packages I don't see the value in duplicating the effort.

In other words, for applications that need this, use Cobra to define your CLI and it's already supported.

**Comment by firelizzard18** [comment]
@tomasaschan How am I supposed to do this with Cobra?

**Comment by macintacos** [comment]
@tomasaschan if this is already supported, is it covered in the User Guide in cobra? I reviewed that and couldn't find anything that resembled what this issue was asking for, but if it's possible in Cobra, at least I'm fine with that (can't speak for others though)

**Comment by tomasaschan** [comment]
I think I might have misremembered the user docs of Cobra; it has features to do this for arguments, but not for flag values. There are, however, hooks for general validation of input, so it can be done there albeit not as nicely as for args.

It can still be done with the current feature set of pflag, using a custom implementation of `Value`, and the fact that @spf13 chose to put the corresponding feature for args in Cobra, rather than in pflag, tells me that sugar for doing it for option values should probably also be outside this package (at least for the time being).

**Comment by shakefu** [comment]
Please don't make us go implement our own func variants for Func/FuncP that are just the permutations of valid String/Int/etc factories. 

This library has already accepted the burden of validation by including such variants as Uint8, etc., validating those numerics match the valid ranges for the Go types. 

If it's an opposition to creating a whole set of `ValidStringP(...)` funcs to initialize the flags bound to validators, I get that. But maybe something like:

```
flags.MarkValidated("someflag", func (value any) error {
    v, ok := any.(int)
    if !ok {
        return errors.New("why wasn't this an int?")
    }
    if v < 10 && v > 0 {
        return nil
    }
    return errors.New("invalid value, must be between 0 and 10")
})
```

Or since strings are the most common needed flag value check:

```
flags.MarkValidStrings("someflag", "stringa", "stringb", "stringc")
```

... I'd be happy to put together a PR if you're open to it.

**Comment by firelizzard18** [comment]
This kind of conversation is one I've seen multiple times on this project. I think ultimately what it comes down to is this: cobra and friends are designed for building command line applications, and they are very opinionated about how you should do that; whereas you see them as helpful libraries and you want to do things that don't fit into their vision of how to build a CLI app. I'm definitely in the second camp - I have no interest in building around the "12 factor app" philosophy (see viper if you don't know what I'm talking about). But ultimately what it means is that cobra/pflags/etc are not built for us. I would love it if someone built a fork or other alternative for pflag that was less opinionated, but "someone" else is the key part there - someone has to be willing to build and maintain that.

**Comment by jerome-diver** [comment]
Yes, it is an observation that seems absolutely correct to me. Perhaps we could also add that since the users concerned do not use the design or the concepts of its tools well, probably at some point there will be a questioning on the documentation to show how we have to comply with the design choices of these modules in cases of exposed use cases?
This could ultimately be a strong point to add this to the existing and already well-supplied documentation. Some examples of alternatives to best use practices with the tools and show what should not be done (if it is right that they can solve specific problems by forcing users to adopt other choices of design pattern of their CLI application with the appropriate standard there). I trust about some solutions by a way or an other. I find my solution, but i can imagine it is not the best use, so... what is the best use for that kind of situation ?

**Comment by shakefu** [comment]
For anyone who comes across this issue later, here's a reference for implementing this for a single string value in a reusable way:

```go
// choiceValue implements the [pflag.Value] interface.
type choiceValue struct {
	value    string
	validate func(string) error
}

// Set sets the value of the choice.
func (f *choiceValue) Set(s string) error {
	err := f.validate(s)
	if err != nil {
		return err
	}

	f.value = s
	return nil
}

// Type returns the type of the choice, which must be "string" for [pflag.FlagSet.GetString].
func (f *choiceValue) Type() string { return "string" }

// String returns the current value of the choice.
func (f *choiceValue) String() string { return f.value }

// StringChoice returns a [choiceValue] that validates the value against a set
// of choices. Only the last value will be used if multiple values are set.
func StringChoice(choices []string) *choiceValue {
	return &choiceValue{
		validate: func(s string) error {
			for _, choice := range choices {
				if s == choice {
					return nil
				}
			}
			return fmt.Errorf("must be one of %v", choices)
		},
	}
}

func main(){
	cmd := &cobra.Command{
		Use:   "choice",
		Short: "Test the choice value.",
		Run: func(cmd *cobra.Command, args []string) {
			item, _ := cmd.Flags().GetString("item")
			fmt.Println("item=", item)
			return nil
		},
	}

	flags := cmd.Flags()
	flags.VarP(StringChoice([]string{
		"uint8", "uint16", "uint32", "uint64",
		"int8", "int16", "int32", "int64",
		"float32", "float64",
	}), "item", "i", "The type of the data line. (default: `float64`).")

	if err := cmd.Execute(); err != nil {
		fmt.Println("error:", err)
	}
}
```

**Comment by fastcat** [comment]
> I think I might have misremembered the user docs of Cobra; it has features to do this for arguments, but not for flag values

I think this is an argument as to why pflag ought to have an enum helper for flag values: Cobra as a library deals with processing positional arguments, and as such includes a helper for enum values there. pflag has/is code for helping process flag arguments, so by the same pattern ought to have a helper for enum values for flags.

It's little surprise to me thus that this is the most thumbs-up'd issue here by roughly an order of magnitude.

**Comment by tomasaschan** [comment]
Ok, I'll reopen this for now, just to signal that the final word on this has not necessarily been said yet.

However, I think this is a feature that would require quite careful API design, if it's to be added at all - I don't particularly like these proposed one here. If I'm the one who'll review a proposal (note: that does not need to be the case), expect a few iterations before we reach something worth taking a stab at implementing. In other words, do not expect a PR for this feature to be merged (by me) just because it technically works...

Furthermore, I'm currently on vacation and will be on parental leave for the rest of the year, so the moments when I have ten consecutive minutes to spend on pflag are, at present, few and far between, just to be honest about expectations here.

**Comment by fastcat** [comment]
> I don't particularly like these proposed one here

When you have a couple of those spare minutes, a bit more on what you don't like about the proposed API(s) here might help discussion.

Just off the top of my head, I can think of some plausible scenarios that might push for different API patterns:
- The set of allowed values is short and known at compile time
  -  A traditional enum
  - A slice-based API as in the original post would work well here
- The set of allowed values is short(-ish), but requires dynamic lookup at runtime
  - E.g. a list of values from a database
  - Validating a candidate value can be done without fetching the whole list, expanding a completion candidate might also be faster than fetching the whole list
- The set of allowed values is large enough (or expensive enough to find) that enumerating it is impractical
  - E.g. every user email in a large organization
  - Validation of a candidate is the main available operation
  - Expanding a completion prefix may be available
- Orthogonal to the above, the value type might be something other than `string`, e.g. a number, or a custom type (`type Foo string`), requiring thought about type parameters or other accommodations

**Comment by jerome-diver** [comment]
> However, I think this is a feature that would require quite careful API design, if it's to be added at all - I don't particularly like these proposed one here. If I'm the one who'll review a proposal (note: that does not need to be the case), expect a few iterations before we reach something worth taking a stab at implementing. In other words, do not expect a PR for this feature to be merged (by me) just because it technically works...


Yes, I think the power to accept or reject isn't reasonably summed up as "I like" or "I don't like." There are certainly reasons. And I think it would be so much more constructive to list them so we can take them into account and do better the way to be clean.
For now, in terms of communication, it looks more like a categorical rejection with a diplomatic air.

Just to be honest, is it possible to become honestly constructive ?
(other than that, since you're talking about it, enjoy your paternity leave.)

**Comment by firelizzard18** [comment]
Please be courteous. Tomas just said that he doesn't like the current proposed API, that's certainly not a categorical rejection. This is an open source project which Tomas is probably not being paid to work on, so you can only expect a limited amount of his time and energy, which has to be spread across the entire project. I agree we kind of need some specifics to develop a more acceptable proposal, but we must be respectful of the maintainers' time and other priorities.

**Comment by jerome-diver** [comment]
Sure, i do not understand why you tell me that because i stay courteous. To ask to be constructive has nothing bad, nothing negative to. Please do not create a drama where there is no factual reason to choose to go that way, and do not imagine dirty thing or make a trial of my intention for free, it is not necessary.
Thank you to stay nice and friendly @firelizzard18, i love you to.

**Comment by tomasaschan** [comment]
To be clear: reopening this issue was intended to signal the opposite of a blanket rejection.

But API design is difficult, and while I can find 20 seconds here and there to post a general answer in an active discussion like this one, I have a much harder time finding time to sit down and think carefully about what a good API would look like. (This comment is written a few words at a time over the course of a whole day...)

I found it better to post something rather than nothing, given the activity on the issue, but also felt I had to make expectations clear: it is likely to take months of calendar time for me to shepherd this feature into the project.

A few properties I care about (but not necessarily an exhaustive list):

* the public API should "rhyme" well with existing features, both in terms of function signatures etc, and in terms of how other extensions of the stdlib feature set are exposed.

* it should be exposed similarly across all flag types (that take values), not just for strings

* the implementation should preferably _not_ be duplicated for every value type

* I _think_ I would prefer to not add a whole other set of function variants - the whole String/Var/P thing is already creating a combinatoric explosion of signatures - and instead use something similar to MarkRequired (or whatever it's called), but this is somewhat in conflict with the first bullet, and one of the reasons I can't just whip up a proposed API that I'll like - I don't have mental space to clearly evaluate these things, and it'll take some experimentation to land something that feels right.

* it should be built in a way that can be extended to support completion script generation (as it is done by Cobra) if this is not already included. I don't know how this works today, so no idea what it'd take.

**Comment by firelizzard18** [comment]
For me it would be valuable to have a flag that maps strings to a constrained set of values. I think there are two different features here:

1. Constraining the values that are valid for a flag without changing the type of the flag variable.
2. Defining a flag that automatically maps from a finite set of strings to a set of target values (not necessarily strings), where the value of the flag variable will be one of the target values (or the default if there's an error).

I agree that (1) is complex. Clearly there is overlap between these, but the primary value (to me) of (2) is expressing that mapping in a way that makes the definition of the flag clear at a glance. Creating a custom flag Value does not achieve that last. I'm thinking of something like this:

<details><summary>enum_flag.go</summary>

```go
var enumFlag int

func init() {
	EnumVar(cmd.Flags(), &enumFlag, "enum", 0, "Example enum flag", map[string]any{
		"foo": 1,
		"bar": 2,
	})
}

func (f *pflag.FlagSet) EnumVar(ptr any, name string, value any, usage string, values map[string]any) {
	rv := reflect.ValueOf(ptr)
	if rv.Kind() != reflect.Pointer {
		panic("ptr must be a pointer")
	} else if rv.IsNil() {
		panic("ptr cannot be nil")
	}
	rv = rv.Elem()
	rv.Set(reflect.ValueOf(value))

	f.Var(&mappedValue{rv, values}, name, usage)
}

type mappedValue struct {
	value  reflect.Value
	lookup map[string]any
}

func (v *mappedValue) Type() string   { return "mapped" }
func (v *mappedValue) String() string { return fmt.Sprint(v.value) }

func (m *mappedValue) Set(s string) error {
	v, ok := m.lookup[s]
	if !ok {
		return fmt.Errorf("%q is not an allowed value", s)
	}
	m.value.Set(reflect.ValueOf(v))
	return nil
}
```

</details>

Obviously the lack of type safety is not great. That could be solved with generics, but EnumVar and friends would need to be functions, not methods (since methods can't be generic), at which point there's not much difference with defining this in a separate package. Still, I would find this definition helpful. It does have the potentially fatal problem that defining `(*FlagSet).Enum(...)` (the version that returns a pointer) isn't really viable due to lack of type safety. Personally that's fine with me since I always use EnumVar, but I can see why that might be a deal breaker. Also passing in a map isn't going to be popular with everyone, but I think it's the best option. I thought of `(*FlagSet).EnumVar(ptr any, name string, value any, usage string, values ...fmt.Stringer)` and building the map on the fly, but I think requiring the map to be stated explicitly makes things a lot clearer.

**Comment by tomasaschan** [comment]
@firelizzard18 I would argue that those are two distinct features, and that while having 1 probably makes it easier to provide good user feedback from the implementation of 2, neither is really a prerequisite of the other. I interpret the original issue description to mean specifically 1, so if 2 is the most useful to you, I think that should be a separate issue.

**Comment by vipcxj** [comment]
I'm surprised such a common feature has been discussed for years without being implemented~
This feature seems simple to implement at first glance, but the deeper you think about it, the more pitfalls you find.
The main problem is how to handle it when there are too many allowed values.
If allowed values are supported, it's not just a matter of adding a validation function; the help output should also reflect the allowed values. If there aren't many allowed values, that's easy and not a problem, but if there are very many, should they still be displayed in full?
Also, should autocomplete automatically support selectable options? If so, it would face the same problem of how to handle it when there are too many options.
But even more frustrating is that it may not get that far — the community is still worrying about how to make the API comfortable and aesthetically pleasing.

**Comment by spenserblack** [comment]
> but if there are very many, should they still be displayed in full?

Generally speaking, if there's enough possible options that the full list is overwhelming, I think it's best for the help message to *not* display all options. Instead, the dev should add an option to the CLI like `--list` or something that would list all available options. For example:

```
example-convert-to-rgb-tool

    --hsl string       convert the provided HSL color to RGB
    --css string       convert the provided CSS color to RGB (call --list-css to see choices)
-L, --list-css         list every CSS color name and exit
```

**Comment by tomasaschan** [comment]
> The main problem is how to handle it when there are too many allowed values.

I don't agree at all that this is the hard part of implementing this feature, although I do agree that it also needs consideration. The easy way out is probably to provide user control both over whether to list all options in the help text, and whether to generate a `--list-options` companion flag (naming etc could be default generated from the flag name, and take inspiration for format from well known and successful CLIs like curl or kubectl). That is yet another thing, though, which then affects the API design.

> But even more frustrating is that it may not get that far — the community is still worrying about how to make the API comfortable and aesthetically pleasing.

In a library like this one, getting the API design right is extremely important, since once there are usages of it in other projects, making changes can be incredibly hard to the point of eliminating the value gain of the change. So far, I've seen one (1) proposed API for this, which to be fair seems to be posted mostly to communicate what the feature should do rather than being a thoroughly thought out API.

@vipcxj I'm sorry if I'm totally misreading something here, but yours comment reads mostly like AI slop to me. What, if anything, are you intending to communicate to further this issue?
